ThirdParty/ecec/src/encrypt.c

```diff
@@ -0,0 +1,292 @@
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+
+// Writes an unsigned 32-bit integer in network byte order.
+static inline void
+ece_write_uint32_be(uint8_t* bytes, uint32_t value) {
+  bytes[0] = (value >> 24) & 0xff;
+  bytes[1] = (value >> 16) & 0xff;
+  bytes[2] = (value >> 8) & 0xff;
+  bytes[3] = value & 0xff;
+}
+
+// Encrypts a plaintext block with optional padding.
+static int
+ece_aes128gcm_encrypt_block(EVP_CIPHER_CTX* ctx, const uint8_t* key,
+                            const uint8_t* iv, const uint8_t* block,
+                            size_t blockLen, const uint8_t* pad, size_t padLen,
+                            uint8_t* record) {
+  int err = ECE_OK;
+
+  // Encrypt the plaintext and padding.
+  if (EVP_EncryptInit_ex(ctx, EVP_aes_128_gcm(), NULL, key, iv) <= 0) {
+    err = ECE_ERROR_ENCRYPT;
+    goto end;
+  }
+  int chunkLen = -1;
+  if (EVP_EncryptUpdate(ctx, record, &chunkLen, block, (int) blockLen) <= 0) {
+    err = ECE_ERROR_ENCRYPT;
+    goto end;
+  }
+  if (EVP_EncryptUpdate(ctx, &record[blockLen], &chunkLen, pad, (int) padLen) <=
+      0) {
+    err = ECE_ERROR_ENCRYPT;
+    goto end;
+  }
+  if (EVP_EncryptFinal_ex(ctx, NULL, &chunkLen) <= 0) {
+    err = ECE_ERROR_ENCRYPT;
+    goto end;
+  }
+
+  // Append the authentication tag.
+  if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, ECE_TAG_LENGTH,
+                          &record[blockLen + padLen]) <= 0) {
+    err = ECE_ERROR_ENCRYPT;
+    goto end;
+  }
+
+end:
+  EVP_CIPHER_CTX_reset(ctx);
+  return err;
+}
+
+// Encrypts a complete message with the given parameters.
+static int
+ece_aes128gcm_encrypt_blocks(EC_KEY* senderPrivKey, EC_KEY* recvPubKey,
+                             const uint8_t* authSecret, size_t authSecretLen,
+                             const uint8_t* salt, size_t saltLen, uint32_t rs,
+                             size_t padLen, const uint8_t* plaintext,
+                             size_t plaintextLen, uint8_t* payload,
+                             size_t* payloadLen) {
+  int err = ECE_OK;
+
+  EVP_CIPHER_CTX* ctx = NULL;
+  uint8_t* pad = NULL;
+
+  if (authSecretLen != ECE_WEBPUSH_AUTH_SECRET_LENGTH) {
+    err = ECE_ERROR_INVALID_AUTH_SECRET;
+    goto end;
+  }
+  if (saltLen != ECE_SALT_LENGTH) {
+    err = ECE_ERROR_INVALID_SALT;
+    goto end;
+  }
+  if (!plaintextLen) {
+    err = ECE_ERROR_ZERO_PLAINTEXT;
+    goto end;
+  }
+
+  // Make sure the payload buffer is large enough to hold the header and
+  // ciphertext.
+  size_t maxRecordEnd =
+    ece_aes128gcm_payload_max_length(rs, padLen, plaintextLen);
+  if (!maxRecordEnd) {
+    err = ECE_ERROR_INVALID_RS;
+    goto end;
+  }
+  if (*payloadLen < maxRecordEnd) {
+    err = ECE_ERROR_OUT_OF_MEMORY;
+    goto end;
+  }
+
+  ctx = EVP_CIPHER_CTX_new();
+  if (!ctx) {
+    err = ECE_ERROR_OUT_OF_MEMORY;
+    goto end;
+  }
+
+  // Allocate enough memory to hold the padding and one-byte padding delimiter.
+  pad = calloc(padLen + 1, sizeof(uint8_t));
```

Comment:
padLen + 1 might overflow and thus creates undefined behaviour in the following if (it might return NULL or not).